ThunkLibs/libOpenCL/libOpenCL_Common.h

```diff
@@ -0,0 +1,229 @@
+#define CL_USE_DEPRECATED_OPENCL_2_0_APIS
+#define CL_USE_DEPRECATED_OPENCL_2_1_APIS
+#define CL_USE_DEPRECATED_OPENCL_2_2_APIS
+#define CL_TARGET_OPENCL_VERSION 300
+#include <CL/cl.h>
+
+namespace FEX::CL {
+  enum class AsyncCallbackType {
+    // Tied to context
+    TYPE_CLCREATECONTEXT,
+    TYPE_CLCREATECONTEXTFROMTYPE,
+    TYPE_CLSETCONTEXTDESTRUCTORCALLBACK,
+    TYPE_CLSETMEMOBJECTDESTRUCTORCALLBACK,
+    TYPE_CLSETPROGRAMRELEASECALLBACK,
+    TYPE_CLSETEVENTCALLBACK,
+    // Tied to cl_progam
+    TYPE_CLBUILDPROGRAM,
+    TYPE_CLCOMPILEPROGRAM,
+    TYPE_CLLINKPROGRAM,
+    // Tied to command_queue
+    TYPE_CLENQUEUENATIVEKERNEL,
+    TYPE_CLENQUEUESVMFREE,
+  };
+
+  union CallbackData {
+    struct _Header {
+      AsyncCallbackType Type;
+#ifdef GUEST_THUNK_LIBRARY
+      uintptr_t cb;
+#else
+      fex_guest_function_ptr cb;
+#endif
+    };
+  };
+
+  struct Type_SingleUserData {
+    CallbackData::_Header Header;
+
+    // User provided data that we need to keep around
+    void *user_data;
+
+    template <AsyncCallbackType Type>
+    static Type_SingleUserData Create(decltype(CallbackData::_Header::cb) a_0, void * a_1) {
+      static_assert(Type == AsyncCallbackType::TYPE_CLCREATECONTEXT ||
+        Type == AsyncCallbackType::TYPE_CLCREATECONTEXTFROMTYPE ||
+        Type == AsyncCallbackType::TYPE_CLSETCONTEXTDESTRUCTORCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETMEMOBJECTDESTRUCTORCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETPROGRAMRELEASECALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETEVENTCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLBUILDPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLCOMPILEPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLLINKPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLENQUEUESVMFREE
+        ,
+        "Mismatch type for handling clProgram");
+
+      FEX::CL::Type_SingleUserData Data {
+        .Header {
+          .Type = Type,
+          .cb = a_0,
+        },
+        .user_data = a_1,
+      };
+
+      return Data;
+    };
+  };
+
+  struct Type_SingleUserDataCopy {
+    CallbackData::_Header Header;
+
+    // User provided data that we need to keep around
+    void *user_data;
+    size_t size;
+    ~Type_SingleUserDataCopy() {
+      free(user_data);
+    }
+
+    template <AsyncCallbackType Type>
+    static std::unique_ptr<Type_SingleUserDataCopy> Create(decltype(CallbackData::_Header::cb) a_0, void * a_1, size_t size) {
+      static_assert(Type == AsyncCallbackType::TYPE_CLENQUEUENATIVEKERNEL,
+        "Mismatch type for handling clProgram");
+
+      void *Memory = malloc(size);
+      memcpy(Memory, a_1, size);
+      return std::make_unique<FEX::CL::Type_SingleUserDataCopy>(FEX::CL::Type_SingleUserDataCopy{
+        .Header {
+          .Type = Type,
+          .cb = a_0,
+        },
+        .user_data = Memory,
+        .size = size,
+      });
+    };
+  };
+
+  union GuestCallbackData {
```

Comment:
Couldn't this be solved more cleanly using `struct GuestCallbackData : CallbackData::_Header`?